### PR TITLE
Apic multi take 2

### PR DESCRIPTION
**Pull request by cjams**

### CODE REVIEW

File: include/hve/arch/intel_x64/irq.h [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+public:
+
+    struct info_t {
+        ::intel_x64::vmcs::value_type vec;
+    };
+
+    using vector_t = decltype(info_t::vec);
+    using irq_t = eapis::intel_x64::irq;
+    using vmcs_t = bfvmm::intel_x64::vmcs;
+    using exit_handler_t = bfvmm::intel_x64::exit_handler;
```
Comment (rianquinn) [inline include/hve/arch/intel_x64/irq.h]: I will likely add these to the base class so that we don't have a ton of these for every class. I'd like to keep these to a minimum (hopefully only the alias in the base class) because otherwise you have to learn an entirely new set of names for each file. I would rather the more esoteric names be fully qualified.

File: include/hve/arch/intel_x64/irq.h [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+    irq &operator=(irq &&) = default;
+
+    irq(const irq &) = delete;
+    irq &operator=(const irq &) = delete;
+
+    /// @endcond
+
+private:
+    exit_handler_t *m_exit_handler;
+    std::array<std::list<handler_t>, 256> m_handlers{};
```
Comment (rianquinn) [inline include/hve/arch/intel_x64/irq.h]: perfect 👍

File: include/hve/arch/intel_x64/irq.h [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+};
+
+} // namespace intel_x64
```
Comment (rianquinn) [inline include/hve/arch/intel_x64/irq.h]: we aren't labeling these in the other files, so for now I would prefer that we stick to that. Now... that being said, if we want to add them, I would merge a PR that did it consistently.

File: include/hve/arch/intel_x64/irq.h [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+namespace reason = ::intel_x64::vmcs::exit_reason::basic_exit_reason;
+
+// -----------------------------------------------------------------------------
+// Definitions
+// -----------------------------------------------------------------------------
+
+///
+/// External interrupt handler
+///
+class EXPORT_EAPIS_HVE irq
```
Comment (rianquinn) [inline include/hve/arch/intel_x64/irq.h]: this should be renamed to "external_interrupt", and then all references to irq should be changed to that. for this class. I would prefer that we stick to the naming the Intel uses so that we do not get lost. New users are going to be reading the manual, so we want them to see the manual, and then see the code an know what they are looking at.

File: include/hve/arch/intel_x64/irq_window.h [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+
+#ifndef IRQWINDOW_HDLR_INTEL_X64_EAPIS_H
```
Comment (rianquinn) [inline include/hve/arch/intel_x64/irq_window.h]: This should be called the "interrupt_window" for the same reason above

File: include/hve/arch/intel_x64/irq.h [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+
+    /// Trap
+    ///
+    /// @expects
+    /// @ensures
+    ///
+    /// Enable external-interrupt exiting. This vcpu will exit each time
+    /// an interrupt arrives during VMX-nonroot operation.
+    ///
+    void trap();
```
Comment (rianquinn) [inline include/hve/arch/intel_x64/irq.h]: enable_trapping

File: include/hve/arch/intel_x64/irq_window.h [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+    void add_handler(handler_t &&d);
+
+    /// Trap
+    ///
+    /// @expects
+    /// @ensures
+    ///
+    /// @param enable enable if true, disable (i.e. passthrough) if false
+    ///
+    void trap(bool enable = true);
```
Comment (rianquinn) [inline include/hve/arch/intel_x64/irq_window.h]: enable_trapping (no param)

File: include/hve/arch/intel_x64/irq_window.h [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+
+    /// Pass-through
+    ///
+    /// @expects
+    /// @ensures
+    ///
+    /// Disable interrupt-window exiting.
+    ///
+    void pass_through();
```
Comment (rianquinn) [inline include/hve/arch/intel_x64/irq_window.h]: not needed. We only support turning the features on.

File: include/hve/arch/intel_x64/irq.h [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+
+    /// Pass Through
+    ///
+    /// @expects
+    /// @ensures
+    ///
+    /// Disable external-interrupt exiting. This vcpu will not exit if
+    /// the vcpu is interrupted during VMX-nonroot operation.
+    ///
+    void pass_through();
```
Comment (rianquinn) [inline include/hve/arch/intel_x64/irq.h]: not needed, we only support turning the features on

File: include/hve/arch/intel_x64/irq_window.h [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+#include <bfvmm/hve/arch/intel_x64/vmcs/vmcs.h>
+#include <bfvmm/hve/arch/intel_x64/exit_handler/exit_handler.h>
+
+// -----------------------------------------------------------------------------
+// Exports
+// -----------------------------------------------------------------------------
+
+#include <bfexports.h>
+
+#ifndef STATIC_EAPIS_HVE
```
Comment (rianquinn) [inline include/hve/arch/intel_x64/irq_window.h]: not needed once I have the base class in place

File: include/hve/arch/intel_x64/irq.h [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+#include <bfvmm/hve/arch/intel_x64/vmcs/vmcs.h>
+#include <bfvmm/hve/arch/intel_x64/exit_handler/exit_handler.h>
+
+// -----------------------------------------------------------------------------
+// Exports
+// -----------------------------------------------------------------------------
+
+#include <bfexports.h>
+
+#ifndef STATIC_EAPIS_HVE
```
Comment (rianquinn) [inline include/hve/arch/intel_x64/irq.h]: not needed once I have the base class in place

File: include/hve/arch/intel_x64/irq_window.h [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+    irq_window(irq_window &&) = default;
+    irq_window &operator=(irq_window &&) = default;
+
+    irq_window(const irq_window &) = delete;
+    irq_window &operator=(const irq_window &) = delete;
+
+    /// @endcond
+
+private:
+    exit_handler_t *m_exit_handler;
```
Comment (rianquinn) [inline include/hve/arch/intel_x64/irq_window.h]: space

File: include/hve/arch/intel_x64/irq.h [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+    irq(irq &&) = default;
+    irq &operator=(irq &&) = default;
+
+    irq(const irq &) = delete;
+    irq &operator=(const irq &) = delete;
+
+    /// @endcond
+
+private:
+    exit_handler_t *m_exit_handler;
```
Comment (rianquinn) [inline include/hve/arch/intel_x64/irq.h]: space

File: include/vcpu/arch/intel_x64/vcpu.h [hunk 14]
```diff
@@ -22,6 +22,8 @@
 #include "../../../hve/arch/intel_x64/msrs.h"
 #include "../../../hve/arch/intel_x64/cpuid.h"
 
+#include "../../../vic/arch/intel_x64/irq_manager.h"
```
Comment (rianquinn) [inline include/vcpu/arch/intel_x64/vcpu.h]: no space

File: include/vcpu/arch/intel_x64/vcpu.h [hunk 15]
```diff
@@ -38,7 +40,28 @@ class vcpu : public bfvmm::intel_x64::vcpu
     ///
     vcpu(vcpuid::type id) :
         bfvmm::intel_x64::vcpu{id}
-    { }
+    {
+        enable_msr_trapping();
+
+        auto handler = this->exit_handler();
```
Comment (rianquinn) [inline include/vcpu/arch/intel_x64/vcpu.h]: Couple things:

- This should be in its own function similar to how we enable the other apis. 
- We should not pass the VMCS to the class as it can get it from the exit handler. 
- The code should look more like this to stay consistent
- Use msrs() instead of directly accessing the member. 

```
m_irqmgr = std::make_unique<eapis::intel_x64::irq_manager>(
    this->exit_handler(), 
    this->msrs()
);
```

**Review by rianquinn — CHANGES_REQUESTED** [review]

File: include/hve/arch/intel_x64/irq_window.h (hunk 8 above)
Comment (cjams) [reply include/hve/arch/intel_x64/irq_window.h]: We need to be able to turn this on and off

File: include/hve/arch/intel_x64/irq.h (hunk 4 above)
Comment (cjams) [reply include/hve/arch/intel_x64/irq.h]: I assume the crs class will be renamed to control_register_access as well?

File: include/vcpu/arch/intel_x64/vcpu.h (hunk 15 above)
Comment (cjams) [reply include/vcpu/arch/intel_x64/vcpu.h]: The vmcs is a private member of exit_handler

File: include/hve/arch/intel_x64/irq_window.h (hunk 8 above)
Comment (rianquinn) [reply include/hve/arch/intel_x64/irq_window.h]: I take that back, this one will be needed (similar to monitor trap) because Intel doesn't disable this for you. So Yeah, this should just be called disable_trapping

File: include/vic/arch/intel_x64/irq_manager.h [hunk 16]
```diff
@@ -0,0 +1,188 @@
...
+{
+
+/// IRQ Manager
+///
+///
+class EXPORT_EAPIS_VIC irq_manager
+{
+public:
+
+    using irq_t = eapis::intel_x64::irq;
```
Comment (rianquinn) [inline include/vic/arch/intel_x64/irq_manager.h]: Most of these can all be removed. You don't need to include eapis::intel_x64 for anything that is already in that namespace, so most of these are redundant, the rest are IMO overkill. Now, that being said, once I have the usual added to the base class, you can use those

File: include/vic/arch/intel_x64/irq_manager.h [hunk 17]
```diff
@@ -0,0 +1,188 @@
...
+
+    /// Constructor
+    ///
+    /// @expects
+    /// @ensures
+    ///
+    irq_manager(
+        gsl::not_null<exit_handler_t *> exit_handler,
+        gsl::not_null<vmcs_t *> vmcs,
+        gsl::not_null<msrs_t *> msrs
```
Comment (rianquinn) [inline include/vic/arch/intel_x64/irq_manager.h]: This is a good example. msrs_t has lost it's context while reading as I don't know if this refers to the msrs in the intrinsics, or the msrs in the eapis. In this case, since the msrs are part of the EAPIs, all you have to do is right "msrs", and then it becomes explicit again as to what your referring to without having to go back and read the alias list.

Comment (rianquinn) [reply include/vic/arch/intel_x64/irq_manager.h]: As a side note, I will add the VMCS alias to the base class, but it will likely be called "vmcs_n" instead of "vmcs_t" as your aliasing a namespace, not a type.

File: include/vic/arch/intel_x64/irq_manager.h [hunk 18]
```diff
@@ -0,0 +1,188 @@
...
+    /// @ensures
+    ///
+    /// If interrupt-window exiting is enabled, an exit occurs on any
+    /// instruction boundary at which the guest is interruptible i.e
+    ///
+    /// guest.rflags.if = 1 &&
+    /// interruptibility_state.blocking_by_sti = 0 &&
+    /// interruptibility_state.blocking_by_mov_ss = 0 &&
+    ///
+    bool handle_irqwin(gsl::not_null<vmcs_t *> vmcs);
```
Comment (rianquinn) [inline include/vic/arch/intel_x64/irq_manager.h]: This should be called "handle_interrupt_window_exit". To reduce space, write the code like this:

```
 bool handle_interrupt_window_exit(
    gsl::not_null<vmcs_t *> vmcs);
```

File: include/vic/arch/intel_x64/irq_manager.h [hunk 19]
```diff
@@ -0,0 +1,188 @@
...
+    /// Handle external-interrupt exit
+    ///
+    /// @expects
+    /// @ensures
+    ///
+    /// Handler for external-interrupt exits
+    ///
+    /// @param vmcs the vmcs for this exit reason
+    ///
+    bool handle_extirq(gsl::not_null<vmcs_t *> vmcs, irq_info_t &info);
```
Comment (rianquinn) [inline include/vic/arch/intel_x64/irq_manager.h]: handle_external_interrupt_exit (its not an IRQ)

File: include/vic/arch/intel_x64/irq_manager.h [hunk 20]
```diff
@@ -0,0 +1,188 @@
...
+    ///
+    /// @expects
+    /// @ensures
+    ///
+    /// Handle an irq, which may come from an external interrupt exit
+    /// or an interrupt window exit.
+    ///
+    /// @param v the vector at which the irq occured
+    ///
+    void handle_irq(vector_t v);
```
Comment (rianquinn) [inline include/vic/arch/intel_x64/irq_manager.h]: This should be renamed to handle_host_interrupt, this way it is explicit that you are handling the case where an interrupt is fired in the host, vs. an exit.

File: include/vic/arch/intel_x64/irq_manager.h [hunk 21]
```diff
@@ -0,0 +1,188 @@
...
+    void init_apic_ctl();
+    void init_x2apic_ctl();
+    void init_save_state();
+    void inject_irq(vector_t v);
+
+    exit_handler_t *m_exit_handler{nullptr};
+    vmcs_t *m_vmcs{nullptr};
+    msrs_t *m_msrs{nullptr};
+
+    std::unique_ptr<irq_t> m_extirq{nullptr};
```
Comment (rianquinn) [inline include/vic/arch/intel_x64/irq_manager.h]: The initializer is not needed for unique_ptr and I actually think it might be a little slower than using the default.

File: include/vic/arch/intel_x64/irq_manager.h [hunk 22]
```diff
@@ -0,0 +1,188 @@
...
+    /// Handle an irq, which may come from an external interrupt exit
+    /// or an interrupt window exit.
+    ///
+    /// @param v the vector at which the irq occured
+    ///
+    void handle_irq(vector_t v);
+
+private:
+
+    /// @cond
```
Comment (rianquinn) [inline include/vic/arch/intel_x64/irq_manager.h]: not needed for private

File: include/vic/arch/intel_x64/irq_manager.h [hunk 23]
```diff
@@ -0,0 +1,188 @@
...
+
+    exit_handler_t *m_exit_handler{nullptr};
+    vmcs_t *m_vmcs{nullptr};
+    msrs_t *m_msrs{nullptr};
+
+    std::unique_ptr<irq_t> m_extirq{nullptr};
+    std::unique_ptr<irq_window_t> m_irqwin{nullptr};
+    std::unique_ptr<lapic_ctl_t> m_lapic_ctl{nullptr};
+
+    // The irq_manager owns the ist, thus once *this is destroyed,
```
Comment (rianquinn) [inline include/vic/arch/intel_x64/irq_manager.h]: Remove comment from header. This should be a better comment to add to the constructor / destructor in the source, but I would use the "note" style comment for editors:

```
// NOTE:
//
// The interrupt_manager owns the "ist", thus once it is ....
```

File: include/vic/arch/intel_x64/irq_manager.h [hunk 24]
```diff
@@ -0,0 +1,188 @@
...
+
+namespace eapis
+{
+namespace intel_x64
+{
+
+/// IRQ Manager
+///
+///
+class EXPORT_EAPIS_VIC irq_manager
```
Comment (rianquinn) [inline include/vic/arch/intel_x64/irq_manager.h]: should be renamed to "interrupt_manager". IRQ has a different meaning.

File: include/vic/arch/intel_x64/irq_manager.h [hunk 25]
```diff
@@ -0,0 +1,188 @@
...
+
+    std::unique_ptr<irq_t> m_extirq{nullptr};
+    std::unique_ptr<irq_window_t> m_irqwin{nullptr};
+    std::unique_ptr<lapic_ctl_t> m_lapic_ctl{nullptr};
+
+    // The irq_manager owns the ist, thus once *this is destroyed,
+    // exceptions will have an invalid stack to work off of
+    std::unique_ptr<gsl::byte[]> m_ist1{nullptr};
+
+    std::deque<vector_t> m_virr;
```
Comment (rianquinn) [inline include/vic/arch/intel_x64/irq_manager.h]: AHHHHHH 😄 

J/K, I know your going to change this. This really should be a `std::list<yourmom, object_allocator<yourmom>>`. Please get in the habit in the future of marking code that you plan to change so that it's not forgotten about. I'm a fan of wrapping such code in a header  / footer comment block in big bold letters "REMOVE ME"

File: include/vic/arch/intel_x64/isr.h [hunk 26]
```diff
@@ -44,43 +44,18 @@
...
-{
-namespace intel_x64
-{
-namespace isr
-{
-
-using exit_hdlr_t = bfvmm::intel_x64::exit_handler;
-void init_vmm_idt(gsl::not_null<exit_hdlr_t *> hdlr);
-
+default_isr(uint64_t vector, uint64_t ec, bool ec_valid,
```
Comment (rianquinn) [inline include/vic/arch/intel_x64/isr.h]: This is not going to format the way you want it to. It should be written as

```
extern "C" EXPORT_SYM void
default_isr(
    uint64_t vector, uint64_t ec, bool ec_valid, uint64_t *regs) noexcept;
```

File: include/vic/arch/intel_x64/isr.h [hunk 27]
```diff
@@ -44,43 +44,18 @@
...
-// -----------------------------------------------------------------------------
-// Interrupt Service Routines
-//
-// The following ISRs are used to populate the IDT. These ISRs are defined
-// in src/vic/arch/intel_x64.isr.asm, and call the default ISR.
-// -----------------------------------------------------------------------------
-
-/// @cond
+extern "C" void
+set_default_isrs(bfvmm::x64::idt *idt,
```
Comment (rianquinn) [inline include/vic/arch/intel_x64/isr.h]: same thing, I would go through your code to verify. We don't want the args being pushed out more than 4 spaces from the left edge, so if you plan to break the args up per line, you have to start with the first arg (i.e. the first arg cannot be on the same line as the function name)

File: include/vic/arch/intel_x64/lapic_ctl.h [hunk 28]
```diff
@@ -0,0 +1,183 @@
...
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+
+#ifndef VIC_LAPIC_INTEL_X64_H
```
Comment (rianquinn) [inline include/vic/arch/intel_x64/lapic_ctl.h]: This doesn't match the file name. It should be: LAPIC_CTL_INTEL_X64_H

File: include/vic/arch/intel_x64/lapic_ctl.h [hunk 29]
```diff
@@ -0,0 +1,183 @@
...
+#pragma warning(disable : 4251)
+#endif
+
+namespace eapis
+{
+namespace intel_x64
+{
+
+namespace intel_msrs = ::intel_x64::msrs;
+namespace intel_lapic = ::intel_x64::lapic;
```
Comment (rianquinn) [inline include/vic/arch/intel_x64/lapic_ctl.h]: Please get rid of these, I would prefer you use "using namespace" in the parts of the code that need this so that it's more explicit to the reader. 

Also.... am I missing something? What is ::intel_x64::lapic? That would suggest it's in the intrinsics, but I don't remember seeing that.

File: include/vic/arch/intel_x64/lapic_ctl.h [hunk 30]
```diff
@@ -0,0 +1,183 @@
...
+
+/// Local APIC base class
+///
+/// This abstract class provides an interface to lapic control operations
+/// that are common to both xAPIC and x2APIC modes.
+///
+struct EXPORT_VIC lapic_ctl
+{
+    using gpa_t = uintptr_t;
+    using field_t = ::intel_x64::msrs::field_type;
```
Comment (rianquinn) [inline include/vic/arch/intel_x64/lapic_ctl.h]: gpa_type, field_type, value_type, and I think I'd love to see a better name for vector_t as I initially read that a std::vector when you are actually referring to the interrupt vector

File: include/vic/arch/intel_x64/lapic_ctl.h [hunk 31]
```diff
@@ -0,0 +1,183 @@
...
+    using value_t = uint64_t;
+    using vector_t = uint64_t;
+
+    enum index : uint32_t { idx0, idx1, idx2, idx3, idx4, idx5, idx6, idx7 };
+    enum lvt_reg : uint32_t { cmci, timer, thermal, perf, lint0, lint1, error };
+    enum count_reg : uint32_t { initial, current };
+    enum reg_op : uint32_t { read, write };
+
+    ///
```
Comment (rianquinn) [inline include/vic/arch/intel_x64/lapic_ctl.h:46]: missing title

File: include/vic/arch/intel_x64/lapic_ctl.h [hunk 32]
```diff
@@ -0,0 +1,183 @@
...
+    ///    operation is allowed.
+    /// @return -1 if the supplied address doesn't map to a valid register or the
+    ///    operation is not allowed.
+    ///
+    /// @param addr - guest physical address of desired register
+    /// @param op - the desired operation (read / write)
+    ///
+    virtual int check_gpa_op(const gpa_t addr, const reg_op op) = 0;
+
+    ///
```
Comment (rianquinn) [inline include/vic/arch/intel_x64/lapic_ctl.h:60]: missing title

File: include/vic/arch/intel_x64/lapic_ctl.h [hunk 33]
```diff
@@ -0,0 +1,183 @@
...
+/// that are common to both xAPIC and x2APIC modes.
+///
+struct EXPORT_VIC lapic_ctl
+{
+    using gpa_t = uintptr_t;
+    using field_t = ::intel_x64::msrs::field_type;
+    using value_t = uint64_t;
+    using vector_t = uint64_t;
+
+    enum index : uint32_t { idx0, idx1, idx2, idx3, idx4, idx5, idx6, idx7 };
```
Comment (rianquinn) [inline include/vic/arch/intel_x64/lapic_ctl.h]: love the syntax here.... going to use that too. My only complaint is the enums should have _t added to them

File: include/vic/arch/intel_x64/lapic_ctl.h [hunk 34]
```diff
@@ -0,0 +1,183 @@
...
+    ///    operation is allowed.
+    /// @return -1 if the supplied address doesn't map to a valid register or the
+    ///    operation is not allowed.
+    ///
+    /// @param addr - MSR address of desired register
+    /// @param op - the desired operation (read / write)
+    ///
+    virtual int check_msr_op(const field_t msr, const reg_op op) = 0;
+
+    virtual value_t read_register(const uint32_t offset) = 0;
```
Comment (rianquinn) [inline include/vic/arch/intel_x64/lapic_ctl.h:74]: I'd like to find a better way to define comment these functions. Since this is a pure virtual base class, I would ditch the comment in favor of your comment blocks like your for "Register reads", and then just wrap this code using `///@cond`

File: include/vic/arch/intel_x64/lapic_ctl.h [hunk 35]
```diff
@@ -0,0 +1,183 @@
...
+    ///
+    virtual int check_gpa_op(const gpa_t addr, const reg_op op) = 0;
+
+    ///
+    /// Check if MSR address is an APIC register and the desired read / write
+    /// operation is allowed.
+    ///
+    /// @return offset if supplied address maps to a valid register and the
+    ///    operation is allowed.
+    /// @return -1 if the supplied address doesn't map to a valid register or the
```
Comment (rianquinn) [inline include/vic/arch/intel_x64/lapic_ctl.h:66]: should this return an exception? We should throw when something bad has happened that should never happen, while tossing an error code like "-1" can be used when something bad happens that will likely happen at some point

File: include/vic/arch/intel_x64/lapic_ctl.h [hunk 36]
```diff
@@ -0,0 +1,183 @@
...
+    /// the vector prior to the call
+    ///
+    virtual bool in_service(const vector_t vec) = 0;
+
+    ///
+    /// Default operations
+    ///
+    /// @cond
+
+    virtual ~lapic_ctl() = default;
```
Comment (rianquinn) [inline include/vic/arch/intel_x64/lapic_ctl.h]: the constructor / destructor should be at the top as usual

Comment (rianquinn) [reply include/vic/arch/intel_x64/lapic_ctl.h]: Also, why "_ctl". Normally when I am reading source code from other projects, this is just called "lapic, xapic and x2apic"

File: include/vic/arch/intel_x64/xapic_ctl.h [hunk 37]
```diff
@@ -0,0 +1,183 @@
...
+/// xAPIC subclass of the lapic abstract base class
+///
+/// This class implements the abstract lapic interface for xapic
+/// mode. It is marked final because it is intended to interact
+/// directly with xapic hardware and thus attempts to avoid the
+/// overhead of virtualized calls as much as possible.
+///
+struct EXPORT_VIC xapic_ctl final : public lapic_ctl
+{
+    using gpa_t = lapic_ctl::gpa_t;
```
Comment (rianquinn) [inline include/vic/arch/intel_x64/xapic_ctl.h]: Not needed. You can use aliases from the base class

File: include/vic/arch/intel_x64/xapic_ctl.h [hunk 38]
```diff
@@ -0,0 +1,183 @@
...
+    ///
+    int check_msr_op(const field_t msr, const reg_op op) noexcept override;
+
+    value_t read_register(const uint32_t offset) noexcept override;
+    void write_register(const uint32_t offset, const value_t val) noexcept override;
+
+    ///
+    /// Register reads
+    ///
+    value_t read_id() noexcept override;
```
Comment (rianquinn) [inline include/vic/arch/intel_x64/xapic_ctl.h:84]: We really should document all of these functions so that the reader can follow what each function is. Might not be a bad idea to grab text from the Intel manual to support the docs here.

File: include/vic/arch/intel_x64/xapic_ctl.h [hunk 39]
```diff
@@ -0,0 +1,183 @@
...
+    xapic_ctl &operator=(xapic_ctl &&) = default;
+
+    xapic_ctl(const xapic_ctl &) = delete;
+    xapic_ctl &operator=(const xapic_ctl &) = delete;
+
+    /// @endcond
+
+
+private:
+    std::array<uint32_t, 128> m_apic_page;
```
Comment (rianquinn) [inline include/vic/arch/intel_x64/xapic_ctl.h]: uint32_t should use the alias

Comment (rianquinn) [inline include/vic/arch/intel_x64/xapic_ctl.h]: space

File: src/hve/arch/intel_x64/irq.cpp [hunk 40]
```diff
@@ -0,0 +1,85 @@
...
+namespace ack_on_exit = exit_ctl::acknowledge_interrupt_on_exit;
+namespace irq_exiting = pin_ctl::external_interrupt_exiting;
+
+/// ---------------------------------------------------------------------------
+/// Helpers
+/// ---------------------------------------------------------------------------
+
+static irq::info_t
+parse_info(gsl::not_null<irq::vmcs_t *> vmcs)
+{ return { exit_irq_info::vector::get() }; }
```
Comment (rianquinn) [inline src/hve/arch/intel_x64/irq.cpp]: why did you add the extra {}?

File: src/hve/arch/intel_x64/irq.cpp [hunk 41]
```diff
@@ -0,0 +1,85 @@
...
+
+/// ---------------------------------------------------------------------------
+/// Implementation
+/// ---------------------------------------------------------------------------
+
+irq::irq(gsl::not_null<exit_handler_t *> exit_handler)
+:
```
Comment (rianquinn) [inline src/hve/arch/intel_x64/irq.cpp]: : should be added to the previous line. The only time you would put : on the next line is if you had to break out the function args

File: src/hve/arch/intel_x64/irq.cpp [hunk 42]
```diff
@@ -0,0 +1,85 @@
...
+{
+    m_exit_handler->add_handler(
+        reason::external_interrupt,
+        handler_delegate_t::create<irq_t, &irq_t::handle>(this)
+    );
+}
+
+void
+irq::add_handler(vector_t vector, handler_t &&d)
+{ m_handlers[vector].push_front(std::move(d)); }
```
Comment (rianquinn) [inline src/hve/arch/intel_x64/irq.cpp]: tisk, tisk. Don't use []

File: src/hve/arch/intel_x64/irq.cpp [hunk 43]
```diff
@@ -0,0 +1,85 @@
...
+
+void
+irq::enable_trapping()
+{
+    ack_on_exit::enable();
```
Comment (rianquinn) [inline src/hve/arch/intel_x64/irq.cpp]: This is a good example of the aliases being used too much here. I would prefer to see the full path. Instead, add a using namespace as the first line here so that you can reduce the size of a single line. I don't want to search for an alias list to read this code

File: src/hve/arch/intel_x64/irq.cpp [hunk 44]
```diff
@@ -0,0 +1,85 @@
...
+    irq_exiting::enable();
+}
+
+bool
+irq::handle(gsl::not_null<vmcs_t *> vmcs)
+{
+    auto info = parse_info(vmcs);
+    for (const auto &d : m_handlers[info.vec]) {
```
Comment (rianquinn) [inline src/hve/arch/intel_x64/irq.cpp]: Please see my implementation of the MSR handlers. You should be using find

File: src/hve/arch/intel_x64/irq.cpp [hunk 45]
```diff
@@ -0,0 +1,85 @@
...
+irq::enable_trapping()
+{
+    ack_on_exit::enable();
+    irq_exiting::enable();
+}
+
+bool
+irq::handle(gsl::not_null<vmcs_t *> vmcs)
+{
+    auto info = parse_info(vmcs);
```
Comment (rianquinn) [inline src/hve/arch/intel_x64/irq.cpp]: If parse info is only used once, leave it in this function so that we can see the process. Also, the way you wrote this causes a copy.

File: src/hve/arch/intel_x64/irq.cpp [hunk 46]
```diff
@@ -0,0 +1,85 @@
...
+    for (const auto &d : m_handlers[info.vec]) {
+        if (d(vmcs, info)) {
```
Comment (rianquinn) [inline src/hve/arch/intel_x64/irq.cpp]: I feel like this should do more after the delegate. Is there any way to make the handler easier for the user by taking care of some common grunt work?

File: src/hve/arch/intel_x64/irq_window.cpp [hunk 47]
```diff
@@ -0,0 +1,66 @@
...
+#include <utility>
+#include <hve/arch/intel_x64/irq_window.h>
+
+namespace eapis
+{
+namespace intel_x64
+{
+
+namespace reason = ::intel_x64::vmcs::exit_reason::basic_exit_reason;
+namespace pri_ctl = ::intel_x64::vmcs::primary_processor_based_vm_execution_controls;
```
Comment (rianquinn) [inline src/hve/arch/intel_x64/irq_window.cpp]: please remove and use "using namespace" in a function if you need to reduce the line length

File: src/hve/arch/intel_x64/irq_window.cpp [hunk 48]
```diff
@@ -0,0 +1,66 @@
...
+void
+irq_window::pass_through()
+{ irq_window_exiting::disable(); }
+
+bool
+irq_window::handle(gsl::not_null<vmcs_t *> vmcs)
+{
+    for (const auto &d : m_handlers) {
+        if (d(vmcs)) {
+            return true;
```
Comment (rianquinn) [inline src/hve/arch/intel_x64/irq_window.cpp]: I'm sure there is more we can do here. For example, the user should not have to enable trapping or disable trapping for the interrupt window. Instead, this class should have the "queue" in it, and if you add to the queue, it enables itself, and if the queue is empty, it automatically disables the interrupt. This way, the user doesn't care. They basically just give the interrupt window class the interrupts that need to be injected, and that's it. The rest of the logic is done by this class, and the only reason you would add a handler is if you had something special you needed to do above and beyond that.

File: src/main/CMakeLists.txt [hunk 49]
```diff
@@ -1,6 +1,6 @@
 #
 # Bareflank Hypervisor
-# Copyright (C) 2015 Assured Information Security, Inc.
+# Copyright (C) 2018 Assured Information Security, Inc.
```
Comment (rianquinn) [inline src/main/CMakeLists.txt:3]: lol.... I will add a note to fix all of these.

File: include/vic/arch/intel_x64/x2apic_ctl.h [hunk 50]
```diff
@@ -0,0 +1,181 @@
...
+    ///
+    virtual bool in_service(const vector_t vec) noexcept override;
+
+    ///
+    /// Default operations
+    ///
+    /// @cond
+
+    virtual ~x2apic_ctl() = default;
+    x2apic_ctl() = default;
```
Comment (rianquinn) [inline include/vic/arch/intel_x64/x2apic_ctl.h]: same thing, constructors and destructors should be at the top.

File: include/vic/arch/intel_x64/xapic_ctl.h [hunk 51]
```diff
@@ -0,0 +1,183 @@
...
+#pragma warning(push)
+#pragma warning(disable : 4251)
+#endif
+
+namespace eapis
+{
+namespace intel_x64
+{
+
+namespace intel_xapic = ::intel_x64::xapic;
```
Comment (rianquinn) [inline include/vic/arch/intel_x64/xapic_ctl.h:29]: not needed

**Review by rianquinn — CHANGES_REQUESTED** [review]
